### PR TITLE
[Publisher Agent] Reconnect if the Broker Dies

### PR DESCRIPTION
**Pull request by harryrackmil**
# Change
We observed a case where the Broker logged a read timeout from the publisher agent and the publisher agent never reconnected, so the publisher agent's data stopped reaching the broker.

I reproduced this situation locally by setting the broker read timeout to 10 nanoseconds and triggering a disconnection, but the publisher agent correctly reconnected to the broker.

It may be the case that something else (like a network outage) caused the publisher agent to not be able to reach the broker for the duration of the read timeout, and meant that the publisher agent never got a close message and therefore never attempted to reconnect.

As a blanket fix, we add a write timeout to the publisher agent so that we'll attempt to reconnect to any broker if it takes more than 10 seconds to write to it. We'll also remove an unnecessary `break` from the reconnect loop - we should really never exit that loop unless we've deregistered the broker.

# Testing
Before making these changes, I ran the publisher agent and broker locally and then stopped the broker container, and observed that the publisher agent just hung infinitely and didn't log any errors or attempt to reconnect.

After these changes, the publisher agent attempts to reconnect every 5 seconds and when I start the broker container up again it resumes publishing normally.

### CODE REVIEW

File: cli/stork-publisher-agent/runner.go [hunk 1]
```diff
@@ -76,10 +76,12 @@ func (r *PublisherAgentRunner[T]) UpdateBrokerConnections() {
 
 	// add or update desired connections
 	for brokerUrl, newAssetIdMap := range newBrokerMap {
-		_, exists := r.brokerMap[brokerUrl]
-		if exists {
+		_, brokerExists := r.brokerMap[brokerUrl]
+		outgoingConnection, outgoingConnectionExists := r.outgoingConnections[brokerUrl]
```
Comment (harryrackmil) [inline cli/stork-publisher-agent/runner.go:80]: got a nil reference error when I was trying to break things and refreshing the registry really frequently - if an outgoing connection object is missing we should create it.

File: cli/stork-publisher-agent/runner.go [hunk 2]
```diff
@@ -147,7 +149,8 @@ func (r *PublisherAgentRunner[T]) RunOutgoingConnection(url BrokerPublishUrl, as
 		conn, _, err := websocket.DefaultDialer.Dial(string(url), headers)
 		if err != nil {
 			r.logger.Error().Err(err).Msgf("Failed to connect to outgoing WebSocket: %v", err)
-			break
+			time.Sleep(r.config.BrokerReconnectDelay)
+			continue
```
Comment (harryrackmil) [inline cli/stork-publisher-agent/runner.go:153]: if we fail to connect to a broker, we should try again in 5 seconds.

File: cli/stork-publisher-agent/websocket.go [hunk 3]
```diff
@@ -312,10 +313,15 @@ func SendWebsocketMsg[T any](conn *websocket.Conn, msgType string, data T, trace
...
-	return sendWebsocketResponse[WebsocketMessage[T]](conn, msg, logger)
+	return sendWebsocketResponse[WebsocketMessage[T]](conn, msg, logger, OutgoingWriteTimeout)
 }
 
-func sendWebsocketResponse[T any](conn *websocket.Conn, msg T, logger zerolog.Logger) error {
+func sendWebsocketResponse[T any](conn *websocket.Conn, msg T, logger zerolog.Logger, writeTimeout time.Duration) error {
+	if writeTimeout.Nanoseconds() > 0 {
+		deadline := time.Now().Add(writeTimeout)
+		_ = conn.SetWriteDeadline(deadline)
+	}
```
Comment (harryrackmil) [inline cli/stork-publisher-agent/websocket.go:323]: set a write deadline so we get a timeout if the broker isn't reachable

File: cli/stork-publisher-agent/signer_test.go [hunk 4]
```diff
@@ -28,6 +28,7 @@ func TestSigner_GetSignedPriceUpdate_Evm(t *testing.T) {
 		"",
 		"",
 		time.Duration(0),
+		time.Duration(0),
```
Comment (harryrackmil) [inline cli/stork-publisher-agent/signer_test.go:31]: fix tests

File: cli/stork-publisher-agent/runner.go [hunk 5]
```diff
@@ -10,16 +10,16 @@ import (
...
-	outgoingConnectionsLock sync.RWMutex
-	signer                  Signer[T]
+	config                      StorkPublisherAgentConfig
+	signatureType               SignatureType
+	logger                      zerolog.Logger
+	ValueUpdateCh               chan ValueUpdate
+	signedPriceBatchCh          chan SignedPriceUpdateBatch[T]
+	registryClient              *RegistryClient
+	assetsByBroker              map[BrokerPublishUrl]map[AssetId]struct{}
+	outgoingConnectionsByBroker map[BrokerPublishUrl]*OutgoingWebsocketConnection[T]
```
Comment (harryrackmil) [inline cli/stork-publisher-agent/runner.go:20]: renamed for clarity - one map stores the connection object and the other stores the assets we want to send through that connection.

File: cli/stork-publisher-agent/runner.go [hunk 6]
```diff
@@ -76,23 +76,23 @@ func (r *PublisherAgentRunner[T]) UpdateBrokerConnections() {
 
 	// add or update desired connections
 	for brokerUrl, newAssetIdMap := range newBrokerMap {
-		_, exists := r.brokerMap[brokerUrl]
-		if exists {
+		outgoingConnection, outgoingConnectionExists := r.outgoingConnectionsByBroker[brokerUrl]
```
Comment (harryrackmil) [inline cli/stork-publisher-agent/runner.go:79]: the old logic here was kind of weird - we really only want to update connections which still exist, regardless of whether we have an existing entry in the broker asset map. 

We now only look at the connection map when deciding whether to create a new connection, and the asset map is basically a configuration for each connection.

**Review by akawalsky — DISMISSED** [review]

**Pull request merged**